jax/_src/lax/parallel.py

```diff
@@ -726,17 +722,96 @@ def all_gather(x, axis_name, *, axis_index_groups=None):
    [[12. 13. 14. 15.]
     [ 4.  5.  6.  7.]]
   """
+  axis_size = psum(1, axis_name, axis_index_groups=axis_index_groups)
+  # The all_gather primitive doesn't work when omni-staging is disabled.
+  if not config.omnistaging_enabled:
+    return _all_gather_via_psum(x, all_gather_dimension=0, axis_name=axis_name,
+                                axis_index_groups=axis_index_groups, axis_size=axis_size)
+
+  def bind(x):
+    return all_gather_p.bind(x, all_gather_dimension=0, axis_name=axis_name,
+                             axis_index_groups=axis_index_groups, axis_size=axis_size)
 
+  return tree_util.tree_map(bind, x)
+
+def _all_gather_via_psum(x, *, all_gather_dimension, axis_name, axis_index_groups, axis_size):
   index = axis_index(axis_name)
   if axis_index_groups is not None:
     indices = np.array(axis_index_groups).flatten()
     axis_index_to_group_index = indices.argsort() % len(axis_index_groups[0])
     index = lax_numpy.array(axis_index_to_group_index)[index]
+  outs = tree_util.tree_map(partial(_expand, all_gather_dimension, axis_size, index), x)
+  return psum(outs, axis_name, axis_index_groups=axis_index_groups)
 
-  axis_size = psum(1, axis_name, axis_index_groups=axis_index_groups)
-
-  return _allgather(x, 0, axis_size, index, axis_name, axis_index_groups)
+def _all_gather_translation_rule(c, x, *, all_gather_dimension, axis_name, axis_index_groups, axis_size, axis_env, platform):
+  # TODO(cjfj): Enable this for TPU also?
```

Comment:
Any reason not to make this part of this change?

Reply:
Only that I haven't run any benchmarks on TPU. If you're confident that it won't introduce a performance regression, I'm happy to enable it on TPU also. Do you happen to know if the TPU op supports any `all_gather_dimension`?

Reply:
TPU supports any all_gather_dimension, I believe you are working around an XLA:GPU bug. XLA:GPU should choose a layout which makes any all_gather_dimension possible.

Reply:
Alternatively, XLA:GPU could use https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/all_gather_decomposer.h#L25-L27

Reply:
I went ahead and did this with [this CL](https://github.com/tensorflow/tensorflow/commit/dd4835f906f925c0465eb1f304d451dba1eff585).

It should not be needed to check either the platform or the all_gather_dimension.

Reply:
Thanks David! I'll enable it on all platforms in a follow-up PR.